Add initial Spark refactor

dfsnow opened the pull request:
This PR represents an initial attempt to refactor the increasingly slow and unwieldy [service-sqoop-iasworld](https://github.com/ccao-data/service-sqoop-iasworld) to Spark. This refactor has numerous benefits:

- Spark is the industry standard for this type of work, while sqoop is deprecated and no longer maintained
- PySpark code is infinitely more readable and maintainable compared to the sqoop code, which is mostly bash
- PySpark extractions tend to be a full order of magnitude faster than sqoop due to the use of SQL predicates to create more discrete extract tasks
  - `ASMT_ALL WHERE taxyr IN (2023, 2024)` via sqoop = 1H30M
  - `ASMT_ALL WHERE taxyr IN (2023, 2024)` via Spark = 14M
- Job definitions are clearer and easier to maintain (via their own dedicated YAML file). They're also far more flexible e.g. only get `cur = 'Y'` values for 2023
- The output of the extract jobs is a zstd-compressed Parquet file per partition, compared to the many files currently created per partition via sqoop. This should make our Athena queries slightly faster
- Making table extracts only via indexed columns (PARID, TAXYR) may actually result in faster queries AND less load on the iasWorld database (compared to Enquire queries, which use a lot of un-indexed columns)

Note that this PR only includes the extract and partitioning parts of the refactor. I figured it would be worthwhile to break the review up into more manageable chunks. I still need to add:

- Upload to S3
- Job validation
- Logging
- Schema assurance/conformity to existing sqoop types

Commit:
Add initial Spark setup

Initial working sqoop image

Add docker-compose setup and pull scripts

Add sample cred files

Add GitLab CI script

Add README

Reorganize

Checkpoint

Working script logic for job creation and data types

Fix extra " in column mappings

Add drivers/ dir for ojdbc

Remove cruft from entrypoint script

Update README.md

Switch curl links to HTTPS

Manually add EPEL rpm file

Add build log artifacts to CI

Fix dockerfile permissions issue

Fix SSH key warning w wpasswordless login

Define CI_IMAGE_TAG for push stage

Remove unneeded RUN from Dockerfile

Setup distributed-in-docker cluster

Fully distributed 3 node cluster working

Update CI script to enable caching

Working mapreduce to local file

Finalize job creation scripts

Add S3 syncing logic

Checkpoint

Fix mappings for large numbers

Update README

Test CI caching fix

Fix sqoop stage build issue

Fix CI caching

Add local Docker build script

Include schema in uploads

Remove orphans on run

Fix entrypoint permissions issue

Remove port 8080 due to conflict with RStudio

Fix permissions issue running entrypoint

Remove orphans on run

Force zstd compression on logs

Update ports for run jobs

Remove incorrect error msg

Fix integer overflows by switching to Long data type

Fix timeout with addition conn string param

Update S3 URI and change log name to ISO 8601

Switch export script to mv instead of sync

AWS currently ingests files then deletes them. This means that using aws
s3 sync will constantly replace the deleted files by copying them from
source (the export server). Switching to mv will delete the files after
they're exported to S3

Remove long-form connection string

Add shell pop info to README

Condense common sqoop options to bash array

Set docker subnet to something we will NEVER use

Fix connection timeout issues for large tables

See https://stackoverflow.com/questions/2327220/oracle-jdbc-intermittent-connection-issue/8122100#8122100

Switch subnet pool

Fix java security bug that caused connection drops

Add network to create-jobs compose file

Want to prevent it from creating its own separate default network

Add inplace flag for sed

Split ASMT_ tables into 3 separate sqoop jobs

This fixes issues with queries on very large tables timing out before
completion

Update table mappings to deal with NUMBER(,) types

Change ASMT_ALL job to only grab CUR = 'Y'

Update Java security params

Remove ASMT_ALL query for now

Split ASMT tables into chunks

Keep ASMT_ALL in queries

Finalize java sec params

Add executable bit to entrypoint script

Switch to uploading only parquet

Ignore backups data folder

Add cronjob to run export at 1 AM each morning

Fix crontab run time (need UTC offset)

Update README

Change incremental type to WEN instead of IASW_ID

iasWorld's IASW_ID column is NOT unique, so it cannot be used for
incremental updates. Switched sqoop import type to lastmodified using
the WEN column

Update README with sqoop resources

Update README with connection error section

Upload sqoop logs to AWS CloudWatch

Fix first-time sqoop import behavior

Need certain workarounds to import correctly:
- Only use num_mappers > 8 on first (big) import, subsequent incremental
  imports should use a single mapper
- The first import should ingest rows where WEN is NULL, subsequent
  imports expect WEN to have a value

Update query to grab NULL WEN values

Remove incremental data loading

iasWorld columns are too dirty to load data incrementally
Lots of NULLs, incorrectly imported values, etc.

Initial hive config finished

Checkpoint

Checkpoint

Checkpoint

Fully working hcat extraction with part/bucketing

Enable bucketing by PARID

Checkpoint

Cleanup all scripts

Fix table mappings

Remove S3 keys before uploading

Update README

Final tweaks and bugfixes

Tweak MR memory settings

Remove additional data nodes from compose setup

Update mem limit for 64GB memory

Move tables list

Remove java json in repo

Add script to get SQL table definitions

Add static Hive table definitions

Convert taxyr to string

Add taxyr partitioning

Wrap col names in backticks

Move taxyr to end up temp table

Fix all lines with TAXYR removed from table def

Remove seq col from CV tables

Remove bucketing (no parid)

Add chunk uploading for logs

Add ability to specify TAXYR for indv. tables

Update crontab schedule

Add SNAPPY compression to tables

Always use log splitting at 5K lines

Update crontab format

Remove temp script used to grab table definitions

Fix num_mappers to number of TAXYRs

Remove empty dirs before S3 upload

Update crontab on README

Fix crontab formatting

Update S3 bucket target

Update crontab schedule

Update cron job schedule and years pulled

Fix typo

Re-add script used to define table schemas

Explicitly set num-mappers for table def script to 1

Ignore java files

Update table definitions after iasw schema update

Keep temporary backup local copy of logs

Update crontab to skip ASMT_HIST and ASMT_ALL

Add exapp table to sqoop job

Add exapp to readme bash syntax

Update SQL schemas for iasWorld updates

Update README.md

Update README

Drop GitLab CI files

Test GHA Docker workflow

Update .env

Disable attestation for Docker images

Add labeling to Docker build

Create CODEOWNERS

Update docker-build.yaml to not tag for PRs

Update table definitions for PERMIT and ENTER

Update README cronjobs with new tables

Cleanup shell scripts and Dockerfile (#4)

* Fix shellcheck errors

* Cleanup all shell scripts with shfmt and shellcheck

* Lint Dockerfile

Add script to dispatch dbt tests after sqoop import (#5)

* Add crawler trigger to run.sh

* Add initial scripts for dispatching GH workflows

* Update example .env file

* Fix small bash quoting issues

* Fix Glue crawler name

* Add logging messages in dispatch script

Hotfix shell script issues (#6)

* Unquote variables in math operations

* Update styles in curl calls

* Fix logging output from dispatch script

Trigger build workflows for `model` on sqoop (#7)

* Trigger build workflows on sqoop

* Rework API_PARAMS generating statement

* Add comment and curly brace

Revert "Trigger build workflows for `model` on sqoop (#7)"

This reverts commit bd596453eaf73761f621bb1feab69807c0b6cd8e.

Add new table schema for taxyr 2024 (#8)

Set maximum number of mappers

Add UTC loaded_at timestamp to iasWorld table pulls (#10)

* Add UTC loaded_at timestamp to iasWorld table pulls

* Update table schema definitions

* Add loaded_at timestamp to extract scripts

Swap alias for table in extract script

`table` is a reserved name / expression and can't be used as an alias

Bump max number of mappers to match core count

Update cronjobs to grab most recent two years

Update sqoop URL to use HTTPS (#11)

* Remove multi-stage build

* Build image on any PR

* Set pipe options for second stage build

* Update sqoop URL to HTTPS

* Revert docker build on non-main branches

Update extracted table definitions (#13)

* Remove bucketing on small tables

* Update table definitions, compression, and buckets

* Update table definitions

* Switch back to snappy compression

* Revert data type changes to tables

* Change only coord columns

* Drop backup table definitions

* Revert comment

Fix missing taxyr cols (#15)

Initial Spark compose setup

Remove deprecated Docker config files

Add working pyspark job

Working memory settings

Add PIN-level partitioning

Working partitioning

Add pyarrow setup

dfsnow committed:
Merge remote-tracking branch 'origin/main' into dfsnow/spark-refactor

dfsnow committed:
Drop sqoop table definitions and files

dfsnow committed:
Drop GitHub workflow

dfsnow committed:
Consolidate docker compose

dfsnow committed:
Setup initial Dockerfile

dfsnow committed:
Add initial pyproject setup

dfsnow committed:
Add docker build GH workflow

dfsnow committed:
Finalize Dockerfile

dfsnow committed:
Drop sqoop scripts

dfsnow committed:
Add linting tooling

dfsnow committed:
Format with ruff

dfsnow committed:
Swap default branch to main

dfsnow committed:
Add LICENSE and metadata

dfsnow committed:
Add yamllint config file

dfsnow committed:
Fix bash linter errors

dfsnow committed:
Reorder env vars

dfsnow committed:
Fix Action shellcheck errors

dfsnow committed:
Add type check to JWT script

dfsnow committed:
Rename scripts to spark dir

dfsnow committed:
Create dedicate scripts dir

dfsnow committed:
Refactor spark test to class

dfsnow committed:
Checkpoint

dfsnow committed:
Split into job and session modules

dfsnow committed:
Restructure target folder

dfsnow committed:
Add PyArrow handling

dfsnow committed:
Add handling for cur

dfsnow committed:
Move predicates reader to utils

dfsnow committed:
Fix filter creation

dfsnow committed:
Cleanup paths and add garbage collection

dfsnow committed:
Checkpoint

dfsnow committed:
Restructure repository

dfsnow committed:
Set worker ports in compose to avoid sqoop conflict

dfsnow committed:
Add submission mechanism for jobs

dfsnow committed:
Add predicate and year handling

dfsnow committed:
Move config options to dedicated file

dfsnow committed:
Move docstrings and set SALES conf

dfsnow committed:
Add a set of default jobs

dfsnow committed:
Set docker user

dfsnow committed:
Lock version dependencies for Python

dfsnow committed:
Add default jobs and submission with yq

dfsnow committed:
Lock Docker deps

dfsnow committed:
Fix bash lint error

dfsnow committed:
Cleanup structure

dfsnow committed:
Fix Docker entrypoints

dfsnow committed:
Add docs

dfsnow committed:
Fix shfmt error

dfsnow committed:
Add initial README

dfsnow committed:
Tweak README

dfsnow committed:
Tweak copy a teeny tiny bit

dfsnow committed:
Fix linting error

dfsnow committed:
Disable shfmt

dfsnow committed:
Re-disable black, isort

dfsnow commented on .env:
This file is just a demo/template. The actual file in the repo is filled out and changes to it are ignored.

dfsnow commented on .github/workflows/build-docker.yaml:
GitHub workflows haven't changed other than bumping the Action versions. Feel free to skip reviewing them.

dfsnow commented on README.md:
This is just a first pass at the README. Will fill it out more as things get added. That said, it should give you a pretty good approximation of what's going on in the repo.

jeancochrane replied:
**[Praise]** Excellent docs, thank you!

dfsnow commented on README.md:
```diff
@@ -1 +1,95 @@
+## Submitting jobs
+
+> [!NOTE]
+> Before attempting to submit jobs to the cluster, first make sure the Spark
+> Docker Compose stack is active by running `docker compose up -d` in the
+> repository.
```
This is a change compared to sqoop, which called `docker compose` via a shell script to spin up and down its cluster. I think the always-on way is preferable since then we can use the Spark web UI to view job history.

dfsnow commented on README.md:
````diff
@@ -1 +1,95 @@
+
+```json
+{
+  "addn": {
````
The "job name" doesn't actually matter. It's not used in any way, just there for reference.

dfsnow commented on config/default_jobs.yaml:
This file defines jobs via YAML, which gets converted to the JSON shown in the README. This format has a lot of benefits:

- We can define "sets" of jobs to run i.e. all weekend jobs or all jobs needed for QC reports
- Jobs only require the table name. Any other fields that aren't defined just use the defaults. This saves lots of space / reduces redundancy.
- We can use YAML templating to change lots of things at once. See `min_year`

dfsnow commented on config/default_predicates.csv:
This file is the big brain bit that makes Spark way faster than sqoop. Here's why:

Both Spark and sqoop extract via JDBC. To do so, the divide up the data that needs to be extracted into partitions, then assign one worker to pull each partition. This works great unless your partitions are drastically different sizes.

For sqoop, we partitioned by `taxyr`, which works well for the most part, but _really_ chugs on tables where a single `taxyr` is huge (like `ASMT_ALL`).

So what can we do to divide up the range of data into smaller chunks? Partition on PIN! This file defines the partition boundaries of roughly equally sized chunks of PINs. Each one of these chunks then gets a worker, which makes extracts super fast.

> [!NOTE]
> One point of uncertainty I have here: I don't account for possibly null values of `parid`, `taxyr`, or `cur`. These aren't supposed to be nullable columns, but :shrug:.

dfsnow commented on docker-compose.yaml:
```diff
@@ -0,0 +1,79 @@
+    environment:
+      - SPARK_MODE=master
+
+      - SPARK_PUBLIC_DNS=${SPARK_PUBLIC_IP}
+      - SPARK_RPC_AUTHENTICATION_ENABLED=no
+      - SPARK_RPC_ENCRYPTION_ENABLED=no
+      - SPARK_LOCAL_STORAGE_ENCRYPTION_ENABLED=no
+      - SPARK_SSL_ENABLED=no
+      - SPARK_USER=spark
+      - IPTS_HOSTNAME
+      - IPTS_PORT
+      - IPTS_SERVICE_NAME
+      - IPTS_USERNAME
+      - IPTS_TABLE
```
These are shared by both containers, but unfortunately I don't think there's a way to dedupe them.

dfsnow commented on run.sh:
This is just a lil example file showing how to submit jobs from YAML. I imagine I'll expand this script so we can use it via cron jobs and the like.

dfsnow commented on src/submit_jobs.py:
This script does all the actual work of submitting and tracking the jobs. It basically: 

1. Turns the JSON into jobs
2. Creates a single Spark session to share across all jobs
3. Runs all the extract jobs, then closes the Spark session to free memory
4. Uses PyArrow to repartition the Spark read output to a ZFS-compressed single file per partition

I tried to comment it thoroughly, so I won't use PR comments too much here. Let me know if anything is unclear.

> [!NOTE]
> P.S. @jeancochrane it would be very helpful if you could review this (and other Python modules) for idiomatic Python, since I'm usually an R guy.

dfsnow commented on src/submit_jobs.py:
```diff
@@ -0,0 +1,116 @@
+    # credentialing, retries, etc.
+    current_datetime = datetime.now().strftime("%Y-%m-%d_%H:%M:%S")
+    session_name = f"iasworld_{current_datetime}"
+    session = SharedSparkSession(
+        app_name=session_name, password_file_path=PATH_IPTS_PASSWORD
+    )
```
Note that you can connect to an existing session by passing the same name, which might be helpful in a future where we want to poll the status of jobs.

dfsnow commented on src/utils/spark.py:
This is the real bulk of the PR. Figuring out all these Spark options and the right abstractions took me awhile. I think I landed on something pretty simple and understandable, but open to suggestions!

dfsnow committed:
Add final_compression attribute

dfsnow commented on .env:
This file is just a demo/template. The actual file in the repo is filled out and changes to it are ignored.

dfsnow commented on .github/workflows/build-docker.yaml:
GitHub workflows haven't changed other than bumping the Action versions. Feel free to skip reviewing them.

dfsnow commented on config/default_jobs.yaml:
This file defines jobs via YAML, which gets converted to the JSON shown in the README. This format has a lot of benefits:

- We can define "sets" of jobs to run i.e. all weekend jobs or all jobs needed for QC reports
- Jobs only require the table name. Any other fields that aren't defined just use the defaults. This saves lots of space / reduces redundancy.
- We can use YAML templating to change lots of things at once. See `min_year`

dfsnow commented on config/default_predicates.csv:
This file is the big brain bit that makes Spark way faster than sqoop. Here's why:

Both Spark and sqoop extract via JDBC. To do so, the divide up the data that needs to be extracted into partitions, then assign one worker to pull each partition. This works great unless your partitions are drastically different sizes.

For sqoop, we partitioned by `taxyr`, which works well for the most part, but _really_ chugs on tables where a single `taxyr` is huge (like `ASMT_ALL`).

So what can we do to divide up the range of data into smaller chunks? Partition on PIN! This file defines the partition boundaries of roughly equally sized chunks of PINs. Each one of these chunks then gets a worker, which makes extracts super fast.

> [!NOTE]
> One point of uncertainty I have here: I don't account for possibly null values of `parid`, `taxyr`, or `cur`. These aren't supposed to be nullable columns, but :shrug:.

dfsnow commented on run.sh:
This is just a lil example file showing how to submit jobs from YAML. I imagine I'll expand this script so we can use it via cron jobs and the like.

dfsnow commented on src/utils/spark.py:
This is the real bulk of the PR. Figuring out all these Spark options and the right abstractions took me awhile. I think I landed on something pretty simple and understandable, but open to suggestions!

I tried to document it thoroughly via docstrings and comments, rather than PR comments, since I want this to be easy to grok next time we come back and work on it.

dfsnow commented on src/utils/spark.py:
```diff
@@ -0,0 +1,223 @@
+            .parquet(self.initial_dir)
+        )
+
+    def repartition(self) -> None:
```
This method is called after _all_ the Spark extract jobs have run. If you try to call it after each job it will crash the server. Spark doesn't like to free memory, neither does PyArrow. They fight to the death and chungus loses.

jeancochrane commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+    mkdir -p /var/lib/apt/lists/partial && \
+    apt-get update && \
+    apt-get install --no-install-recommends -y wget=1.21.3-1+b2 jq=1.6-2.1 && \
+    wget https://github.com/mikefarah/yq/releases/latest/download/yq_linux_amd64 -q -O /usr/bin/yq && \
```
**[Question, non-blocking]** What's your reasoning for pinning `jq` and `wget` to specific versions but not pinning `yq`?

dfsnow replied:
Being dumb and forgetful mostly. Fixed in ca7e883!

jeancochrane commented on README.md:
```diff
@@ -1 +1,95 @@
+  Default is `1999`.
+- `max_year (optional)` - Maximum tax year (inclusive) to extract from the
+  table. To extract a single year, set `min_year` and `max_year` to the same
+  value. Default is the current year.
```
**[Question, non-blocking]** Based on the single-year instructions I take it the `min_year` parameter is also inclusive, yeah?

dfsnow replied:
Indeed! Clarified in b249e95.

jeancochrane commented on run.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# Example command to execute test jobs in the Spark cluster
+# First run `docker compose up` in the repo root
```
**[Nitpick, non-blocking]** Hopefully readers will understand what you're getting at, but in case we want to be more explicit:

```suggestion
# First run `docker compose up -d` in the repo root
```

jeancochrane commented on src/utils/helpers.py:
```diff
@@ -0,0 +1,95 @@
+
+
+def construct_predicates(
+    csv_lines: list[list[str]], taxyr: int | list[int] | None = None
```
**[Question, non-blocking]** As `src.submit_jobs.main` is currently written, `taxyr` will always be either `NoneType` or `list[int]`, right? Are we preserving the option of `int` in anticipation of a future where that might be more convenient than a one-element list?

dfsnow replied:
Just leftovers from when I initially passed a `taxyr` value directly from the job definition. I think it's worth leaving in just to be defensive.

dfsnow replied:
Changed! See https://github.com/ccao-data/service-spark-iasworld/pull/1#discussion_r1714303698.

jeancochrane commented on src/utils/helpers.py:
```diff
@@ -0,0 +1,95 @@
+    taxyr_list = [taxyr] if isinstance(taxyr, int) else taxyr
+    if taxyr_list:
+        predicates = [
+            f"parid >= '{start}' AND parid <= '{end}' AND taxyr = {year}"
```
**[Question, non-blocking]** Is `taxyr` stored as an int in the database that this predicate is referencing, or are single quotes just optional in this case?

dfsnow replied:
`taxyr` is indeed always an int in all iasWorld tables that I know of.

jeancochrane commented on src/utils/helpers.py:
```diff
@@ -0,0 +1,95 @@
+        A list of start and end PARIDs used to divide a table. The third value
+        is the rough number of PARIDs in the batch.
+    """
+    pred_path = Path(path)
+    with open(pred_path.resolve().as_posix(), mode="r") as file:
+        csv_reader = csv.reader(file)
+        next(csv_reader)
+        csv_lines = [row for row in csv_reader]
+
+    return csv_lines
```
**[Thought, non-blocking]** Since the predicate reading logic in this function seems to be a pretty simple operation, and its output only seems to be used by `construct_predicates`, I wonder if we could simplify things by just having `construct_predicates` accept a path to a predicates file and take responsibility for reading the predicates itself. I don't have strong feelings about it, just keeping an eye out for ways to keep the logic as simple as we can.

dfsnow replied:
I originally did this to avoid reading the CSV on every loop, but I reworked predicates entirely in fd7ffdc. Now this function expects a CSV path that can be specified per job which contains a column, start, and end value. These values are used to create arbitrary SQL BETWEEN expressions. I also removed the interaction with taxyr for simplicity.

jeancochrane commented on src/utils/spark.py:
```diff
@@ -0,0 +1,223 @@
+        taxyr: int | list[int] | None,
+        cur: str | list[str] | None,
+        predicates: list[str] | None,
+        use_partitions: bool,
```
**[Thought, non-blocking]** Not a big deal since it's one extra param that directly maps onto a job config param, but it seems like the value of `use_partitions` could be derivable from the values of `cur` and `taxyr` in case we want to simplify things by getting rid of a parameter.

dfsnow replied:
Good call, fixed in 3f3c5fb! Partitions are now just automatically created if either `cur` or `taxyr` is set.

jeancochrane commented on src/utils/spark.py:
```diff
@@ -0,0 +1,223 @@
+        if self.use_partitions and (
+            (self.taxyr is None) != (self.cur is None)
+        ):
+            raise ValueError(
+                (
+                    f"Error for table {self.table_name}: "
+                    "both 'taxyr' and 'cur' must be set if one is set."
+                )
+            )
```
**[Question, non-blocking]** Given that both `cur` and `taxyr` have default values that are set in `main`, is it possible to configure job parameters that accidentally set one of `taxyr` or `cur` but not the other? Or is this just defensive programming for a future where we might be instantiating `SparkJob` differently (e.g. via an API)? If it's the latter, is there a good reason we don't want to allow a caller to set one or the other of these params but not both?

dfsnow replied:
This was just defensiveness on my part, but I redid partitioning in 3f3c5fb such that partitions are just automatically created if `taxyr` or `cur` are set, since there _are_ tables where there's a `cur` column but no `taxyr` (SALES).

jeancochrane commented on src/utils/helpers.py:
```diff
@@ -0,0 +1,95 @@
+    return table_name
+
+
+def load_job_config(args: Namespace) -> dict:
```
**[Suggestion, non-blocking]** I have a preference for functions that explicitly name (and type hint) their arguments rather than accepting one object with an arbitrary number of attributes, since the latter design forces readers to read the implementation of the function in order to know which attributes the function ends up using. It's not a big deal though, and I think people disagree on this!

dfsnow replied:
Good call! Fixed in dcce2ac.

jeancochrane commented on src/submit_jobs.py:
```diff
@@ -0,0 +1,116 @@
+
+    # Use pyarrow to condense the many small Parquet files created by the reads
+    # into single files per Hive partition
+    for config in jobs:
+        config.repartition()
```
**[Nitpick, non-blocking]** I find `config` to be a bit of a confusing name in this case since it was being used earlier in the function to describe a different object. Is there a reason not to just use `spark_job` again as we do on line `100`?

```suggestion
    for spark_job in jobs:
        spark_job.repartition()
```

dfsnow replied:
Fixed in 79cfdd5! I also renamed some other stuff for clarity.

jeancochrane commented on src/submit_jobs.py:
```diff
@@ -0,0 +1,116 @@
+
+        # Run the Spark read job. Each job will create n_predicates X n_years
+        # files in the target/initial/ dir, assuming predicates are enabled
+        spark_job.read()
```
**[Question, non-blocking]** If I'm understanding correctly, this blocks until the full matrix of predicate reads is complete, is that right? If so, it's making me wonder if there's possibly even more performance that we could eke out by parallelizing the jobs themselves too, such that workers within a job could move onto the next job if all predicate reads have been picked up but not yet completed. Highly speculative, so not something I expect us to investigate now!

dfsnow replied:
Yep that's right. That's a good idea! Added in fa34a59. It doesn't seem to have much impact on performance, but it also doesn't seem to hurt so 🤷 .

jeancochrane commented on src/utils/spark.py:
```diff
@@ -0,0 +1,223 @@
+        if self.taxyr is None:
+            return None
+
+        if isinstance(self.taxyr, list):
+            filter = f"taxyr IN ({', '.join(map(str, self.taxyr))})"
+        else:
+            filter = f"taxyr = {self.taxyr}"
```
**[Question, non-blocking]** Maybe i'm missing something, but aren't we already filtering by `taxyr` via the SQL expressions that comprise the predicates matrix?

dfsnow replied:
I simplified the filter setup a bit in 4090211. Like partitions, the filter is now just constructed from any passed `taxyr` and `cur` values. I also removed the interaction of predicates and `taxyr`.

jeancochrane reviewed (approved):
Really awesome work here! I have no major complaints, just a few questions and small suggestions below.

From a big-picture design perspective, one thing I notice is that the `SparkJob` class is currently designed around the assumption that `cur`, `parid`, and `taxyr` are the only options for our filters/predicates/partitions. If we ever need to change that assumption, I expect we'll need to rewrite most of the `SparkJob` methods to accommodate new or changed options. I think this is probably fine, since we can always refactor the class to make the filter/predicate/partition options more general, but it's one way in which the abstraction as it's currently designed isn't maximally flexible.

I also found the distinction between filters, predicates, and partitions to be a bit confusing, since they all seem to be different kinds of logic conditions, but I think I finally grok it after reading the code. Let me know if this matches your understanding:

* **Predicates** are logic conditions that are applied to read queries _during_ query execution against the iasWorld database, and are used to construct a matrix of mutually exclusive queries that can run in parallel. 
    * Predicates are only enabled if `use_predicates` is set, in which case they are always `parid` and `taxyr`. Users can override the `parid` predicates by editing the predicates CSV file, and can override the `taxyr` predicates by setting `min_year` and/or `max_year` in the job config.
* **Filters** are logic conditions that are applied to the result set _after_ query execution against the iasWorld database.
    * Filters are effectively only enabled if `use_partitions` is set, since otherwise the `main` function will set the `years` and `cur` variables to `None` rather than reading them from the job config. When enabled, filters are always `cur` and `taxyr`. Users can override the values for these filters by setting `min_year`, `max_year`, and/or `cur` in the job config.
* **Partitions** are logic conditions that are applied to the result set after filtering in order to split up the output files and index the data.
    * Partitions are only enabled if `use_partitions` is set, in which case they are always `taxyr` and `cur`. Users can override the values for these filters by setting `min_year`, `max_year`, and/or `cur` in the job config.

There are probably some opportunities for simplification here, even outside of an effort to make these conditions more flexible -- for example, the fact that the filters are enabled by the `use_predicates` job config option is a bit confusing, as is the way that  the script uses `taxyr` as both a predicate and a filter. Still, I don't think any of this is super important to worry about right now!

dfsnow committed:
Pin YQ version in Dockerfile

dfsnow committed:
Clarify min_year inclusivity

dfsnow committed:
Update run.sh

Co-authored-by: Jean Cochrane <jeancochrane@users.noreply.github.com>

dfsnow committed:
Rework predicates file and helpers

dfsnow committed:
Create partitions from TAXYR and CUR if set

dfsnow committed:
Rename functions and vars for clarity

dfsnow committed:
Make args to job_defs function explicit

dfsnow committed:
Simplify filter creation

dfsnow committed:
Reconfigure predicates, cleanup documentation

dfsnow committed:
Change default predicates file extension

dfsnow commented:
Thanks for the great feedback @jeancochrane! I managed to simplify things quite a bit I think. Ready for one more round of review.

> From a big-picture design perspective, one thing I notice is that the `SparkJob` class is currently designed around the assumption that `cur`, `parid`, and `taxyr` are the only options for our filters/predicates/partitions. If we ever need to change that assumption, I expect we'll need to rewrite most of the `SparkJob` methods to accommodate new or changed options. I think this is probably fine, since we can always refactor the class to make the filter/predicate/partition options more general, but it's one way in which the abstraction as it's currently designed isn't maximally flexible.

I'm okay with this. I think it's really unlikely that we'll need to change it, and if we do then it shouldn't be too hard to rewrite. It's really simple (now) and we should keep it that way. That said, I rewrote the predicates part to be column agnostic, see below.

> I also found the distinction between filters, predicates, and partitions to be a bit confusing, since they all seem to be different kinds of logic conditions, but I think I finally grok it after reading the code.

Your understanding of these items is mostly correct, but I agree that they were really confusing. So, I scrapped the existing logic for something much simpler:

- **Predicates** are now entirely separate from `parid` and `taxyr`. They're just arbitrary SQL BETWEEN expressions loaded from a file for each job. This means we could load an arbitrary set of predicates per table. For example `SALES` could be divided up by `saledt` instead of `parid`.
- **Filters** and **partitions** are just automatically enabled as long as either `taxyr` or `cur` is not null. For example, `min_year=2021`, `max_year=2023` and `cur='Y'` will yield all data between 2021 and 2023 (inclusive), where `cur='Y'`, output to files partitioned by both `taxyr` and `cur`. If both the year values and `cur` are falsey, no filtering or partitioning is performed (we just grab the whole table).
 
> There are probably some opportunities for simplification here, even outside of an effort to make these conditions more flexible -- for example, the fact that the filters are enabled by the `use_predicates` job config option is a bit confusing, as is the way that the script uses `taxyr` as both a predicate and a filter. Still, I don't think any of this is super important to worry about right now!

Done! I think I've simplified quite a bit. Let me know what you think!

dfsnow committed:
Add job parallelization

dfsnow committed:
Add helper function for parallelism

dfsnow committed:
Add job description for use in UI

dfsnow committed:
Remove cur from job definitions where not present

dfsnow committed:
Increase number of parallel jobs

Commit:
Finalize job definitions

jeancochrane commented on README.md:
```diff
@@ -1 +1,115 @@
+  table. Defaults to `1999`.
+- `max_year (optional)` - Maximum tax year (inclusive) to extract from the
+  table. To extract a single year, set `min_year` and `max_year` to the same
+  value. Defaults to the current year.
```
**[Suggestion, non-blocking]** On first read, I wasn't clear whether `null` would affect these two columns in the same way it affects `cur` and `predicates_path`:

```suggestion
- `min_year (optional)` - Minimum tax year (inclusive) to extract from the
  table. Set to `null` in a job definition to ignore this column when filtering
  and partitioning. Defaults to `1999`.
- `max_year (optional)` - Maximum tax year (inclusive) to extract from the
  table. To extract a single year, set `min_year` and `max_year` to the same
  value. Set to `null` in a job definition to ignore this column when filtering
  and partitioning. Defaults to the current year.
```

jeancochrane commented on src/submit_jobs.py:
```diff
@@ -0,0 +1,126 @@
+
+        # Run the Spark read job. Each job will create N predicates
+        # files in the target/initial/ dir, assuming predicates are enabled
+        jobs.append(spark_job)
```
**[Nitpick, non-blocking]** I think this comment is out of date, right?

```suggestion
        # Register the job so that we can run them all in parallel once
        # they have all been registered. Each job will create N predicates
        # files in the target/initial/ dir, assuming predicates are enabled
        jobs.append(spark_job)
```

jeancochrane commented on src/utils/helpers.py:
```diff
@@ -0,0 +1,62 @@
+
+    Args:
+        json_file: String path to
+            either `json_file` or `json_string` value, but not both.
```
**[Question, non-blocking]** Are we missing a couple of lines here? I can't quite scan what this sentence is trying to say, and the `Args` section seems to be missing an entry for `json_string`.

dfsnow replied:
Woops! Good catch. Fixed in 015737f620165f8d49a053e438ccb183790d7e4f.

jeancochrane commented on src/utils/helpers.py:
```diff
@@ -0,0 +1,62 @@
+    return table_name
+
+
+def load_job_definitions(json_file: str, json_string: str) -> dict:
```
**[Nitpick, non-blocking]** Since these parameters are loaded from `argparse` arguments with `nargs` set to `?`, I think they should fall back to a [default](https://docs.python.org/3/library/argparse.html#default) of `NoneType`, right?

```suggestion
def load_job_definitions(json_file: str | None, json_string: str | None) -> dict:
```

jeancochrane commented on README.md:
```diff
@@ -1 +1,115 @@
+
+  Predicates are defined via a file of SQL statements in the `config/`
+  directory, then passed to each table job via a file path.
+- **Filters** are logic conditions applied to a query as part of the initial
+  read. They're effectively a SQL WHERE clause, and apply before each read.
```
**[Question, non-blocking]** I originally thought "as part of the initial read" meant filters would be applied _during_ the read, but I think you mean they're applied _after_ the read, is that right? That is, filters are applied to the dataframe that results from the Spark read, whereas predicates are used to chunk the read itself?

dfsnow replied:
Filters are actually composed into each SQL query made to the DB using "predicate pushdown." That means they're executed as a part of each read, rather than after. I clarified this a bit in e41f8e608989a1ad373069bbfa2464a30dee0bd0.

jeancochrane commented on config/default_predicates.sql:
**[Question, non-blocking]** Nice refactor of this file! So did you determine that the additional `taxyr` predicates weren't helpful for speeding things up and reducing database load? Given that `taxyr` is a primary key in most tables I assumed it would be important for improving query performance when we only want a subset of years, but if it doesn't make a difference then I agree that this new format makes things much simpler.

dfsnow replied:
I did some testing and it seems like the addition of `taxyr` in the predicates has basically no effect outside of spawning more job tasks. The only exception is the super huge tables like `ASMT_ALL`, which each group of PINs in a BETWEEN predicate could be hundreds of thousands of rows. I'm fine with the performance <> simplicity tradeoff here though. If it becomes an issue we can make a dedicated predicates file/handler for certain tables.

jeancochrane reviewed (approved):
Awesome refactor! I agree that the new design is much simpler. A couple questions below, but nothing that I consider blocking. Feel free to re-request my review if you end up changing anything major, and otherwise feel free to merge once you've resolved our conversations to your liking.

dfsnow committed:
Update README.md

Co-authored-by: Jean Cochrane <jeancochrane@users.noreply.github.com>

dfsnow committed:
Update src/submit_jobs.py

Co-authored-by: Jean Cochrane <jeancochrane@users.noreply.github.com>

dfsnow committed:
Update src/utils/helpers.py

Co-authored-by: Jean Cochrane <jeancochrane@users.noreply.github.com>

Commit:
Clarify filter push down

Commit:
Add IS NULL default predicate

Commit:
Fix missing docstring
